core/src/main/scala/scalaz/EphemeralStream.scala

```diff
@@ -202,13 +202,19 @@ sealed abstract class EphemeralStreamInstances {
     }
     override def zipWithR[A, B, C](fa: EphemeralStream[A], fb: EphemeralStream[B])(f: (Option[A], B) => C) =
       zipWithL(fb, fa)((b, a) => f(a, b))
-    def traverseImpl[G[_], A, B](fa: EphemeralStream[A])(f: A => G[B])(implicit G: Applicative[G]): G[EphemeralStream[B]] = {
-      val seed: G[EphemeralStream[B]] = G.point(EphemeralStream[B]())
+    
+    def traverseImpl[F[_], A, B](fa: EphemeralStream[A])(f: A => F[B])(implicit F: Applicative[F]): F[EphemeralStream[B]] = {
+      val revOpt: Maybe[F[EphemeralStream[B]]] =
+        F.unfoldrOpt[EphemeralStream[A], B, EphemeralStream[B]](fa)(_ match {
+          case a ##:: as => Maybe.just((f(a), as))
+          case emptyEphemeralStream => Maybe.empty
+        })(Reducer.ReverseEphemeralStreamReducer[B])
 
-      fa.foldRight(seed) {
-        (x , ys) => G.apply2(f(x), ys)((b, bs) => EphemeralStream.cons(b, bs))
-      }
+      val rev: F[EphemeralStream[B]] = revOpt getOrElse F.point(emptyEphemeralStream)
+
+      F.map(rev)(_.reverse)
```

Comment:
It looks like this implementation will not work for Infinite streams. I think the previous implementation was, provided the `Applicative` was lazy enough.
I guess we probably need both a `lazyTraverse` and a `strictTraverse`...

Reply:
There's a test showing it works for an infinite stream (with a short-circuiting implementation of the `Applicative`'s `unfoldrOpt`).

Reply:
`reverse` works for infinite stream??

Reply:
If the `reverse` is reached, the stream was either not infinite, or traversal terminated early and there's no stream inside that `F[EphemeralStream[B]]`.

Reply:
If `List` was used for that unfolding, as I suggest in another comment, it would be obvious that the reversal is safe.

Reply:
yeah.
I guess what I am concerned about is that the identity law (`traverse Identity = Identity`) is not preserved by this implementation for infinite streams. In haskell I can do
```haskell
λ> fmap (take 10) $ traverse Identity [1..]
Identity [1,2,3,4,5,6,7,8,9,10]
```

Reply:
right, looks like `apply2`/`ap2`/`ap` is too strict for this to work.

Reply:
Actually, it could have worked before and I'm having a [déjà vu](https://github.com/scalaz/scalaz/issues/1515#issuecomment-350547944) of thinking about this. 😃

This `Apply#unfoldrOpt` is useful for cases when appending any more `F[A]`s to an `F[B]` won't change its value.

Turns out it doesn't help when those further `F[A]`s cannot be ignored completely, but their effect part can. Examples:
 - when `F` = `Id`, the effect part of all `F[A]`s can be ignored;
 - when `F` = `Writer[Boolean @@ Conjunction, ?]`, once `false` is encountered, the effect part of all subsequent `F[A]`s can be ignored.

(I can't think of any other cases than `Writer` and `Id` (≈ `Writer[Unit, ?]`).)

I think these cases could be supported at the expense of complicating `unfoldrOpt` signature¹. I would keep it out of this PR, though.

---

¹ Such as the following, changing the return value to `Maybe[F[B]] \/ (Maybe[F[B]], S, F[A] => A)`:

```scala
def unfoldrOpt[S, A, B](seed: S)(f: S => Maybe[(F[A], S)])(implicit R: Reducer[A, B]): Maybe[F[B]] \/ (Maybe[F[B]], S, F[A] => A)
```

where

 - return value of `-\/` means that the seed has been exhausted;
 - return value of `\/-` means that there's some leftover seed, but the remaining effects can be ignored by the returned function `F[A] => A`.

Implementation for `Id` would immediately return `\/-((Maybe.empty, seed, identity[A]))`.